File: backend/services/xueqiu.py
"""Snowball (Xueqiu) market data integration for real-time order validation."""

from __future__ import annotations

import logging
import os
import time
from datetime import datetime
from typing import Any, Dict, List

import requests

logger = logging.getLogger(__name__)


class XueqiuMarketDataError(RuntimeError):
    """Base exception for Xueqiu market data failures."""


def _parse_cookie_string(cookie_string: str) -> Dict[str, str]:
    cookies: Dict[str, str] = {}
    for part in cookie_string.split(";"):
        if "=" not in part:
            continue
        name, value = part.split("=", 1)
        cookies[name.strip()] = value.strip()
    return cookies


def _format_symbol(symbol: str, market: str) -> str:
    sym = symbol.upper().strip()
    if "." in sym:
        return sym

    if market == "HK":
        return sym.zfill(5)

    if market == "CN":
        if sym.startswith("6"):
            return f"SH{sym}"
        elif sym.startswith("8"):
            return f"BJ{sym}"
        else:
            return f"SZ{sym}"

    return sym


class XueqiuMarketData:
    """Client wrapper for Snowball minute level K-line endpoints."""

    BASE_URL = "https://stock.xueqiu.com/v5/stock/chart/kline.json"

    def __init__(self) -> None:
        self.session = requests.Session()
        self._session_configured = False
        self._cookie_string: str | None = None
        self._has_user_cookie = False
        self._has_env_cookie = False
        self._env_cookie_invalid = False

    def _configure_session(self) -> None:
        if self._session_configured:
            return

        headers = {
            "Accept": "application/json, text/plain, */*",
            "Accept-Language": "zh-CN,zh;q=0.9",
            "User-Agent": os.getenv(
                "XUEQIU_USER_AGENT",
                "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) "
                "AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0 Safari/537.36",
            ),
            "Referer": os.getenv("XUEQIU_REFERER", "https://xueqiu.com"),
            "Connection": "keep-alive",
        }
        self.session.headers.clear()
        self.session.headers.update(headers)

        cookies: Dict[str, str] = {}
        env_cookie_present = False

        if self._cookie_string:
            cookies.update(_parse_cookie_string(self._cookie_string))
        elif not self._env_cookie_invalid:
            cookie_env = os.getenv("XUEQIU_COOKIES") or os.getenv("XUEQIU_COOKIE")
            if cookie_env:
                cookies.update(_parse_cookie_string(cookie_env))
                env_cookie_present = True

        if not self._env_cookie_invalid:
            token_env_mapping = {
                "xq_a_token": "XUEQIU_TOKEN",
                "xq_r_token": "XUEQIU_R_TOKEN",
                "xq_id_token": "XUEQIU_ID_TOKEN",
            }
            for cookie_name, env_name in token_env_mapping.items():
                env_value = os.getenv(env_name)
                if env_value and cookie_name not in cookies:
                    cookies[cookie_name] = env_value
                    env_cookie_present = True

        self.session.cookies.clear()
        if cookies:
            self.session.cookies.update(cookies)

        self._has_env_cookie = env_cookie_present

        self._session_configured = True

    def set_cookie_string(self, cookie_string: str | None) -> None:
        self._cookie_string = cookie_string.strip() if cookie_string else None
        self._has_user_cookie = bool(self._cookie_string)
        self._session_configured = False
        self._configure_session()
        if self._cookie_string:
            logger.info("Xueqiu cookie string updated")
        else:
            logger.info("Xueqiu cookie string cleared")

    def has_user_cookie(self) -> bool:
        self._configure_session()
        return self._has_user_cookie

    def has_any_cookie(self) -> bool:
        self._configure_session()
        return self._has_user_cookie or self._has_env_cookie

    def _mark_cookie_invalid(self) -> None:
        if self._has_user_cookie:
            logger.warning("Provided Snowball cookie string appears invalid; clearing it.")
            self._cookie_string = None
            self._has_user_cookie = False
        else:
            logger.warning("Environment Snowball cookie appears invalid; disabling it for this process.")
            self._env_cookie_invalid = True
            self._has_env_cookie = False
        self._session_configured = False

    def get_kline_data(
        self,
        symbol: str,
        market: str,
        period: str = "1m",
        count: int = 100,
    ) -> Dict[str, Any]:
        self._configure_session()

        formatted_symbol = _format_symbol(symbol, market)
        params = {
            "symbol": formatted_symbol,
            "begin": int(time.time() * 1000),
            "period": period,
            "type": "before",
            "count": -abs(count),
            "indicator": "kline",
        }

        try:
            response = self.session.get(self.BASE_URL, params=params, timeout=10)
            response.raise_for_status()
            data = response.json()
        except Exception as exc:
            raise XueqiuMarketDataError(
                f"Failed to fetch kline for {formatted_symbol}: {exc}"
            ) from exc

        if not data:
            raise XueqiuMarketDataError(f"Empty response when requesting {formatted_symbol}")

        if "error_code" in data and data.get("error_code"):
            error_code = data.get("error_code")
            error_desc = data.get("error_description") or data.get("error_msg") or "Unknown error"
            self._mark_cookie_invalid()
            raise XueqiuMarketDataError(
                f"Snowball API error ({error_code}): {error_desc}. Cookie may be invalid or expired."
            )

        if "data" not in data:
            self._mark_cookie_invalid()
            raise XueqiuMarketDataError(
                f"Invalid kline payload for {formatted_symbol}; cookie may be invalid."
            )

        return data

    def get_latest_price(self, symbol: str, market: str) -> float:
        kline_data = self.get_kline_data(symbol, market, period="1m", count=1)
        data = kline_data.get("data", {})
        items: List[List[Any]] = data.get("item", [])  # type: ignore[assignment]
        columns: List[str] = data.get("column", [])  # type: ignore[assignment]

        if not items or not columns or "close" not in columns:
            self._mark_cookie_invalid()
            raise XueqiuMarketDataError(
                "Snowball response missing price information; cookie may be invalid or expired."
            )

        latest_item = items[0]
        close_index = columns.index("close")
        try:
            return float(latest_item[close_index])
        except (TypeError, ValueError, IndexError) as exc:
            self._mark_cookie_invalid()
            raise XueqiuMarketDataError("Invalid close price data") from exc

    def parse_kline_data(self, raw_data: Dict[str, Any]) -> List[Dict[str, Any]]:
        data = raw_data.get("data", {})
        columns: List[str] = data.get("column", [])  # type: ignore[assignment]
        items: List[List[Any]] = data.get("item", [])  # type: ignore[assignment]

        if not columns or not items:
            return []

        column_map = {col: idx for idx, col in enumerate(columns)}
        parsed: List[Dict[str, Any]] = []

        for item in items:
            kline_record: Dict[str, Any] = {}

            if "timestamp" in column_map:
                try:
                    timestamp = int(item[column_map["timestamp"]])
                except (TypeError, ValueError, IndexError):
                    timestamp = None
                if timestamp:
                    kline_record["timestamp"] = timestamp
                    kline_record["datetime"] = datetime.fromtimestamp(timestamp / 1000)

            for field in ("open", "high", "low", "close", "volume", "amount"):
                idx = column_map.get(field)
                if idx is None or idx >= len(item):
                    continue
                value = item[idx]
                kline_record[field] = float(value) if value is not None else None

            for field in ("chg", "percent"):
                idx = column_map.get(field)
                if idx is None or idx >= len(item):
                    continue
                value = item[idx]
                kline_record[field] = float(value) if value is not None else None

            parsed.append(kline_record)

        return parsed

    def get_market_status(self, symbol: str, market: str) -> Dict[str, Any]:
        current_time = datetime.now()
        hour = current_time.hour

        if market == "US":
            trading = 21 <= hour <= 23 or 0 <= hour <= 4
        elif market == "HK":
            trading = 9 <= hour < 16
        else:
            trading = 9 <= hour < 15

        status = "TRADING" if trading else "CLOSED"

        return {
            "symbol": symbol,
            "market": market,
            "market_status": status,
            "timestamp": int(time.time() * 1000),
            "current_time": current_time.isoformat(),
        }


xueqiu_client = XueqiuMarketData()


def get_last_price_from_xueqiu(symbol: str, market: str) -> float:
    price = xueqiu_client.get_latest_price(symbol, market)
    if price <= 0:
        raise XueqiuMarketDataError(f"Invalid latest price for {symbol} ({market})")
    return price


def get_kline_data_from_xueqiu(
    symbol: str,
    market: str,
    period: str = "1m",
    count: int = 100,
) -> List[Dict[str, Any]]:
    raw_data = xueqiu_client.get_kline_data(symbol, market, period, count)
    parsed = xueqiu_client.parse_kline_data(raw_data)
    if not parsed:
        raise XueqiuMarketDataError(f"Failed to parse kline data for {symbol} ({market})")
    return parsed


def set_xueqiu_cookie_string(cookie_string: str | None) -> None:
    xueqiu_client.set_cookie_string(cookie_string)

File: hono-backend/src/xueqiu.ts
const BASE_URL = 'https://stock.xueqiu.com/v5/stock/chart/kline.json'

type MarketType = 'US' | 'HK' | 'CN'

type CookieSource = 'user' | 'env' | null

const TOKEN_ENV_MAPPING: Record<string, string> = {
  xq_a_token: 'XUEQIU_TOKEN',
  xq_r_token: 'XUEQIU_R_TOKEN',
  xq_id_token: 'XUEQIU_ID_TOKEN',
}

// 全局cookie状态 - 简化版本
let globalCookie: string | null = null

// 导出全局cookie变量以便调试
export const getGlobalCookie = () => globalCookie

export class XueqiuMarketDataError extends Error {
  constructor(message: string) {
    super(message)
    this.name = 'XueqiuMarketDataError'
  }
}

const DEFAULT_HEADERS: Record<string, string> = {
  Accept: 'application/json, text/plain, */*',
  'Accept-Language': 'zh-CN,zh;q=0.9',
  'User-Agent':
    process.env.XUEQIU_USER_AGENT ??
    'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0 Safari/537.36',
  Referer: process.env.XUEQIU_REFERER ?? 'https://xueqiu.com',
  Connection: 'keep-alive',
}

const parseCookieString = (cookieString: string): Map<string, string> => {
  const cookies = new Map<string, string>()
  cookieString.split(';').forEach((part) => {
    const trimmed = part.trim()
    if (!trimmed) return
    const eqIndex = trimmed.indexOf('=')
    if (eqIndex <= 0) return
    const name = trimmed.slice(0, eqIndex).trim()
    const value = trimmed.slice(eqIndex + 1).trim()
    if (name) {
      cookies.set(name, value)
    }
  })
  return cookies
}

const mergeCookies = (target: Map<string, string>, source: string) => {
  parseCookieString(source).forEach((value, key) => {
    if (!target.has(key)) {
      target.set(key, value)
    }
  })
}

const collectCookies = () => {
  const header = globalCookie?.trim() || undefined
  const cookieSource: CookieSource = globalCookie ? 'user' : null
  
  return {
    header,
    cookieSource,
    envCookiePresent: false,
    hasCookies: !!globalCookie,
  }
}

const markCookieInvalid = (source: CookieSource) => {
  if (source === 'user') {
    console.warn('Provided Snowball cookie string appears invalid; clearing it.')
    globalCookie = null
  }
}

const formatHkSymbol = (value: string) => {
  let core = value
  if (core.endsWith('.HK')) {
    core = core.slice(0, -3)
  }
  if (core.startsWith('HK')) {
    core = core.slice(2)
  }

  const digits = core.replace(/[^0-9]/g, '')
  if (!digits) {
    return core
  }

  return digits.padStart(5, '0')
}

const formatCnSymbol = (value: string) => {
  let core = value
  if (core.endsWith('.CN')) {
    core = core.slice(0, -3)
  }

  let prefix: 'SH' | 'SZ' | 'BJ' | null = null
  const knownPrefixes = ['SH', 'SZ', 'BJ'] as const
  for (const candidate of knownPrefixes) {
    if (core.startsWith(candidate)) {
      prefix = candidate
      core = core.slice(candidate.length)
      break
    }
  }

  const digits = core.replace(/[^0-9]/g, '')
  if (!prefix) {
    if (digits.startsWith('6')) prefix = 'SH'
    else if (digits.startsWith('8')) prefix = 'BJ'
    else prefix = 'SZ'
  }

  prefix = prefix ?? 'SZ'

  if (!digits) {
    return `${prefix}${core}`
  }

  return `${prefix}${digits}`
}

const formatSymbol = (symbol: string, market: MarketType) => {
  const upper = symbol.toUpperCase().trim()
  if (!upper) return upper

  if (upper.includes('.')) {
    const [base, suffix] = upper.split('.', 2)
    if (suffix === 'HK') {
      return formatHkSymbol(base)
    }
    if (suffix === 'US') {
      return base
    }
    if (suffix === 'CN') {
      return formatCnSymbol(base)
    }
    return upper
  }

  if (market === 'HK') {
    return formatHkSymbol(upper)
  }

  if (market === 'CN') {
    return formatCnSymbol(upper)
  }

  return upper
}

export const hasUserCookie = () => Boolean(globalCookie)

export const hasAnyCookie = () => Boolean(globalCookie)

export const setCookieString = (cookieString: string | null | undefined) => {
  const sanitized = cookieString?.trim()
  globalCookie = sanitized ? sanitized : null
  console.log(`[setCookieString] Global cookie ${globalCookie ? 'set' : 'cleared'}, length: ${globalCookie?.length || 0}`)
}

export const clearCookieString = () => {
  globalCookie = null
  console.log(`[clearCookieString] Global cookie cleared`)
}

interface FetchKlineOptions {
  period?: string
  count?: number
  skipCookieCheck?: boolean
}

interface ParsedKlineRecord {
  timestamp?: number
  datetime?: Date
  open?: number
  high?: number
  low?: number
  close?: number
  volume?: number
  amount?: number
  chg?: number
  percent?: number
}

const fetchKline = async (
  symbol: string,
  market: MarketType,
  options: FetchKlineOptions = {},
) => {
  console.log(`[fetchKline] Starting for ${symbol} (${market})`)
  
  if (!options.skipCookieCheck && !hasAnyCookie()) {
    console.error(`[fetchKline] No cookie available`)
    throw new XueqiuMarketDataError(
      'Snowball cookie not configured. Please set the cookie string before requesting market data.',
    )
  }

  const { header: cookieHeader, cookieSource, hasCookies } = collectCookies()
  console.log(`[fetchKline] Cookie source: ${cookieSource}, has cookies: ${hasCookies}`)
  console.log(`[fetchKline] Cookie header length: ${cookieHeader?.length || 0}`)

  const formattedSymbol = formatSymbol(symbol, market)
  console.log(`[fetchKline] Formatted symbol: ${formattedSymbol}`)

  const params = new URLSearchParams({
    symbol: formattedSymbol,
    begin: Math.floor(Date.now()).toString(),
    period: options.period ?? '1m',
    type: 'before',
    count: (-Math.abs(options.count ?? 100)).toString(),
    indicator: 'kline',
  })

  const headers: Record<string, string> = { ...DEFAULT_HEADERS }
  if (cookieHeader) {
    headers.Cookie = cookieHeader
  }

  const url = `${BASE_URL}?${params.toString()}`
  console.log(`[fetchKline] Fetching URL: ${url}`)
  const headersForLog = { ...headers }
  if (headersForLog.Cookie) {
    headersForLog.Cookie = `[redacted length=${headersForLog.Cookie.length}]`
  }
  console.log(`[fetchKline] Headers:`, headersForLog)

  let response: Response
  try {
    response = await fetch(url, {
      headers,
      method: 'GET',
    })
    console.log(`[fetchKline] Response status: ${response.status}`)
    console.log(`[fetchKline] Response headers:`, Object.fromEntries(response.headers.entries()))
  } catch (error) {
    console.error(`[fetchKline] Fetch error:`, error)
    throw new XueqiuMarketDataError(`Failed to call Snowball API: ${(error as Error).message}`)
  }

  if (!response.ok) {
    console.error(`[fetchKline] HTTP error: ${response.status}`)
    if (response.status === 401 || response.status === 403) {
      markCookieInvalid(cookieSource)
    }
    throw new XueqiuMarketDataError(`Snowball API HTTP error: ${response.status}`)
  }

  let data: any
  try {
    const text = await response.text()
    console.log(`[fetchKline] Response text length: ${text.length}`)
    console.log(`[fetchKline] Response text preview: ${text.substring(0, 200)}...`)
    data = JSON.parse(text)
    console.log(`[fetchKline] Parsed JSON keys:`, Object.keys(data))
  } catch (error) {
    console.error(`[fetchKline] JSON parse error:`, error)
    throw new XueqiuMarketDataError('Snowball API returned invalid JSON payload')
  }

  if (!data) {
    console.error(`[fetchKline] Empty data object`)
    throw new XueqiuMarketDataError('Snowball API returned an empty payload')
  }

  if (typeof data.error_code !== 'undefined' && data.error_code) {
    console.error(`[fetchKline] API error: ${data.error_code} - ${data.error_description || data.error_msg}`)
    markCookieInvalid(cookieSource)
    const description =
      data.error_description ?? data.error_msg ?? 'Unknown Snowball API error'
    throw new XueqiuMarketDataError(`Snowball API error (${data.error_code}): ${description}`)
  }

  if (!data.data) {
    console.error(`[fetchKline] Missing data field in response`)
    console.log(`[fetchKline] Full response:`, JSON.stringify(data, null, 2))
    markCookieInvalid(cookieSource)
    throw new XueqiuMarketDataError('Snowball API response missing data field; cookie may be invalid.')
  }

  console.log(`[fetchKline] Success! Returning data`)
  return data
}


export const getLatestPrice = async (
  symbol: string,
  market: MarketType,
  options?: FetchKlineOptions,
): Promise<number> => {
  console.log(`[getLatestPrice] Starting price fetch for ${symbol} (${market})`)
  console.log(`[getLatestPrice] Cookie status - hasAny: ${hasAnyCookie()}, hasUser: ${hasUserCookie()}`)
  console.log(`[getLatestPrice] Global cookie:`, 
    globalCookie ? `[length: ${globalCookie.length}]` : 'null')
  
  try {
    const payload = await fetchKline(symbol, market, { ...options, count: 1 })
    console.log(`[getLatestPrice] fetchKline successful, payload:`, JSON.stringify(payload, null, 2))
    
    const data = payload.data ?? {}
    console.log(`[getLatestPrice] data:`, JSON.stringify(data, null, 2))
    
    // 尝试不同的数据格式
    if (data.quote && data.quote.current) {
      const price = Number(data.quote.current)
      if (Number.isFinite(price) && price > 0) {
        console.log(`[getLatestPrice] Got price from data.quote.current:`, price)
        return price
      }
    }
    
    if (data.last_close) {
      const price = Number(data.last_close)
      if (Number.isFinite(price) && price > 0) {
        console.log(`[getLatestPrice] Got price from data.last_close:`, price)
        return price
      }
    }
    
    // 原始逻辑
    const columns: string[] = Array.isArray(data.column) ? data.column : []
    const items: unknown[] = Array.isArray(data.item) ? data.item : []
    
    console.log(`[getLatestPrice] columns:`, columns)
    console.log(`[getLatestPrice] items count:`, items.length)
    console.log(`[getLatestPrice] first item:`, items[0])

    if (!columns.length || !items.length) {
      console.error(`[getLatestPrice] Missing data - columns.length: ${columns.length}, items.length: ${items.length}`)
      throw new XueqiuMarketDataError('Snowball response missing price information.')
    }

    const closeIndex = columns.indexOf('close')
    console.log(`[getLatestPrice] closeIndex:`, closeIndex)
    
    if (closeIndex < 0) {
      console.error(`[getLatestPrice] No 'close' column found in:`, columns)
      throw new XueqiuMarketDataError('Snowball response missing close column.')
    }

    const latest = Array.isArray(items[0]) ? (items[0] as Array<unknown>) : []
    const value = latest[closeIndex]
    
    console.log(`[getLatestPrice] latest array:`, latest)
    console.log(`[getLatestPrice] close value at index ${closeIndex}:`, value)

    const price = typeof value === 'number' ? value : Number(value)
    console.log(`[getLatestPrice] parsed price:`, price)
    
    if (!Number.isFinite(price) || price <= 0) {
      console.error(`[getLatestPrice] Invalid price: ${price} (isFinite: ${Number.isFinite(price)}, > 0: ${price > 0})`)
      throw new XueqiuMarketDataError('Snowball returned an invalid latest price.')
    }
    
    console.log(`[getLatestPrice] Success! Returning price: ${price}`)
    return price
  } catch (error) {
    console.error(`[getLatestPrice] Error:`, error)
    
    // 如果是因为cookie问题导致失败，尝试清除cookie
    if (error instanceof XueqiuMarketDataError && error.message.includes('cookie')) {
      console.log(`[getLatestPrice] Clearing cookie due to error`)
      clearCookieString()
    }
    
    throw error
  }
}

export const parseKlineData = (rawData: any): ParsedKlineRecord[] => {
  const data = rawData?.data ?? {}
  const columns: string[] = Array.isArray(data.column) ? data.column : []
  const items: unknown[] = Array.isArray(data.item) ? data.item : []

  if (!columns.length || !items.length) {
    return []
  }

  const columnIndex = new Map<string, number>()
  columns.forEach((column, index) => {
    columnIndex.set(column, index)
  })

  return items.map((itemRaw) => {
    const item = Array.isArray(itemRaw) ? (itemRaw as Array<unknown>) : []
    const record: ParsedKlineRecord = {}

    const timestampIndex = columnIndex.get('timestamp')
    if (timestampIndex !== undefined && timestampIndex < item.length) {
      const tsValue = item[timestampIndex]
      const tsNumber = typeof tsValue === 'number' ? tsValue : Number(tsValue ?? 0)
      if (Number.isFinite(tsNumber) && tsNumber > 0) {
        record.timestamp = tsNumber
        record.datetime = new Date(tsNumber)
      }
    }

    for (const field of ['open', 'high', 'low', 'close', 'volume', 'amount', 'chg', 'percent'] as const) {
      const fieldIndex = columnIndex.get(field)
      if (fieldIndex === undefined || fieldIndex >= item.length) continue
      const rawValue = item[fieldIndex]
      if (rawValue === null || rawValue === undefined) continue
      const num = typeof rawValue === 'number' ? rawValue : Number(rawValue)
      if (Number.isFinite(num)) {
        ;(record as any)[field] = num
      }
    }

    return record
  })
}

export const getKlineData = async (
  symbol: string,
  market: MarketType,
  options?: FetchKlineOptions,
): Promise<ParsedKlineRecord[]> => {
  const raw = await fetchKline(symbol, market, options)
  const parsed = parseKlineData(raw)
  if (!parsed.length) {
    throw new XueqiuMarketDataError('Failed to parse Snowball kline data.')
  }
  return parsed
}

export const getMarketStatus = (symbol: string, market: MarketType) => {
  const now = new Date()
  const hour = now.getHours()

  let trading: boolean
  if (market === 'US') {
    trading = (hour >= 21 && hour <= 23) || (hour >= 0 && hour <= 4)
  } else if (market === 'HK') {
    trading = hour >= 9 && hour < 16
  } else {
    trading = hour >= 9 && hour < 15
  }

  return {
    symbol,
    market,
    market_status: trading ? 'TRADING' : 'CLOSED',
    timestamp: Date.now(),
    current_time: now.toISOString(),
  }
}

export type { ParsedKlineRecord, MarketType }
